bundles/core/org.openhab.core.compat1x/src/main/java/org/openhab/io/multimedia/tts/internal/TTSServiceFactory.java

```diff
@@ -0,0 +1,120 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.io.multimedia.tts.internal;
+
+import java.net.URL;
+import java.util.Collection;
+import java.util.Dictionary;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.Set;
+
+import org.openhab.core.compat1x.internal.CompatibilityActivator;
+import org.openhab.io.multimedia.tts.TTSService;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class listens for services that implement the old tts service interface and registers
+ * an according service for each under the new interface.
+ * 
+ * @author Tobias Bräutigam - Initial contribution and API (copied from ActionServiceFactory)
+ */
+public class TTSServiceFactory {
+	private static final Logger logger = LoggerFactory.getLogger(TTSServiceFactory.class);
+	
+	private Map<String, ServiceRegistration<org.eclipse.smarthome.io.multimedia.tts.TTSService>> delegates = new HashMap<>();
+	private BundleContext context;
+	
+	private Set<TTSService> ttsServices = new HashSet<>();
+	
+	public void activate(BundleContext context) {
+		this.context = context;
+		for(TTSService service : ttsServices) {
+			registerDelegateService(service);
+		}
+	}
+	
+	public void deactivate() {
+		for(ServiceRegistration<org.eclipse.smarthome.io.multimedia.tts.TTSService> serviceReg : delegates.values()) {
+			serviceReg.unregister();
+		}
+		delegates.clear();
+		this.context = null;
+	}
+	
+	public void addTTSService(TTSService service) {
+		if(context!=null) {
+			registerDelegateService(service);			
+		} else {
+			ttsServices.add(service);
+		}
+	}
+
+	public void removeActionService(TTSService service) {
+		if(context!=null) {
+			unregisterDelegateService(service);
+		}
+	}
+
+	private void registerDelegateService(TTSService ttsService) {
+		if(!delegates.containsKey(ttsService.getClass().getName())) {
+			TTSServiceDelegate service = new TTSServiceDelegate(ttsService);
+			Dictionary<String, Object> props = new Hashtable<String, Object>();
+			try {
+				String filterString = null;//"("+Constants.BUNDLE_SYMBOLICNAME+"="+ttsService.getClass().getPackage().getName()+")";
+				Collection<ServiceReference<TTSService>> refs = CompatibilityActivator.getContext().getServiceReferences(TTSService.class,filterString);
+				if (refs!=null && refs.size()>0) {
+					for (ServiceReference<TTSService> ref : refs) {
+						Enumeration<URL> e = ref.getBundle().findEntries("src/main/java", "*.java", true);
+						boolean found = false;
+						while (e.hasMoreElements()) {
```

Comment:
You are right, this is indeed not a good solution!
What you would need instead of the TTSService itself is its ServiceReference (on which you can read the properties). This might be helpful: http://stackoverflow.com/questions/7401969/osgi-declarative-service-obtain-servicereference-in-bind-method

So in the addTTSService method, you should take a ServiceReference as an argument. I have never tried this out myself, but I would hope that it works this way.